sound/soc/sof/intel/hda-mlink.c

```diff
@@ -255,6 +255,46 @@ static bool hdaml_link_check_interrupt(u32 __iomem *lctl)
 	return val & AZX_ML_LCTL_INTSTS;
 }
 
+static int hdaml_wait_bit(void __iomem *base, int offset, u32 mask, u32 target)
+{
+	int timeout = HDAML_POLL_DELAY_RETRY;
+	u32 reg_read;
+
+	do {
+		reg_read = readl(base + offset);
+		if ((reg_read & mask) == target)
+			return 0;
+
+		timeout--;
+		usleep_range(HDAML_POLL_DELAY_MIN_US,
+			     HDAML_POLL_DELAY_MIN_US + HDAML_POLL_DELAY_SLACK_US);
+	} while (timeout != 0);
+
+	return -EAGAIN;
```

Comment:
I think this is an open coded `readl_poll_timeout()` (no need for the `read_poll_timeout()` as you don't sleep before the first read).

Reply:
it's the same argument: simple code that anyone can understand v. optimized macro that requires specific knowledge.

Reply:
right, sure. However readl_poll_timeout() is not that complicated to grasp and to be honest, not much need to know the implementation details. It does what it is supposed to.

The only argument to not use it is that we specifically want to open implement and not reuse code already done, which is a valid one, I guess.

Reply:
no change for now, I added an explanation in the commit message